### DISCUSSION TITLE
Example writing a complex structure to parquet.

### DISCUSSION THREAD

**Discussion by tghunt** [apache/arrow-rs#4558] [General]
Hello all,

Does anyone know of an example writing complex records to parquet files using the parquet crate; Where the record would be something like "field1, field 2, field3, Vec<struct<string, Vec<struct<string, string>>>>".  The record example reflects how data would exists in BigQuery after ingestion.

**Comment by tustvold** [comment]
Have you considered using the arrow APIs? These will handle the record shredding for you and so should be easier to get started with

**Reply by tghunt** [reply]
I did look into arrow as well and gravitated to the RecordBuilder.  However, that does not seem to work for this layout as all arrays in each column have to be the same length.  

The file will be ingested into BigQuery and should allow for the creation of the nested data per record there.  I have seen some theoretical discussion that its possible in this [link](https://arrow.apache.org/blog/2022/10/17/arrow-parquet-encoding-part-3/), however, translating that into example code has proven difficult since I am new to the arrow and parquet.

I have done this with avro which was considerably easier to work with than arrow and parquet, however, writing to avro file is doable but not a solution for the size of data.

**Reply by tustvold** [reply]
Perhaps some of the below might be helpful

https://docs.rs/arrow-array/latest/arrow_array/builder/struct.GenericListBuilder.html#method.append_value

https://docs.rs/arrow-array/latest/arrow_array/builder/index.html#nested-usage

https://docs.rs/arrow-array/latest/arrow_array/builder/index.html#custom-builders

https://docs.rs/arrow-array/latest/arrow_array/array/struct.GenericListArray.html#method.new

**Reply by tghunt** [reply]
I have toiled over this without success and I keep ending up with "all columns in a record batch must have the same length" from the RecordBatch.  While different size nested structures can be created in arrow, it does not seem like nested structures are supported when creating a RecordBatch.

**Reply by tustvold** [reply]
Perhaps you could share your code and I can advise, RecordBatch definitely supports nested lists

**Reply by tghunt** [reply]
![bq_schema](https://github.com/apache/arrow-rs/assets/739514/14add537-6f98-4391-97e2-bc329e067d70)
![bq_nested_data](https://github.com/apache/arrow-rs/assets/739514/59d11ca8-bbee-44db-9a78-dc29e9a0a5e5)
[main.txt](https://github.com/apache/arrow-rs/files/12156060/main.txt)

I have attached an image of the BigQuery schema, and the output along with the latest version of the code that gets me part way there.  If I add to the vec![] for the StructArrays I get the error message I noted earlier.  Note that in the code I have a working vec! line and a non-working vec! line.

Basically, I am not sure how (if possible) to get the StructArray deeper.

**Note:  I the image shows an equal number of items in each record columns.  In practice this is not the case and I should have had three in the last record column.

**Reply by tustvold** [reply]
```
use arrow::array::{
    Int32Builder, MapBuilder, MapFieldNames, StringBuilder, TimestampNanosecondBuilder,
};
use arrow::record_batch::RecordBatch;
use arrow::util::pretty::pretty_format_batches;
use std::sync::Arc;

struct RowBuilder {
    time: TimestampNanosecondBuilder,
    obj1: StringBuilder,
    many_name_value_pairs: MapBuilder<StringBuilder, Int32Builder>,
}

impl RowBuilder {
    fn new() -> Self {
        Self {
            time: Default::default(),
            obj1: Default::default(),
            many_name_value_pairs: MapBuilder::new(
                Some(MapFieldNames {
                    entry: "entries".to_string(),
                    key: "name".to_string(),
                    value: "value".to_string(),
                }),
                Default::default(),
                Default::default(),
            ),
        }
    }

    fn finish(&mut self) -> RecordBatch {
        RecordBatch::try_from_iter(vec![
            ("time", Arc::new(self.time.finish()) as _),
            ("obj1", Arc::new(self.obj1.finish()) as _),
            (
                "many_name_value_pairs",
                Arc::new(self.many_name_value_pairs.finish()) as _,
            ),
        ])
        .unwrap()
    }
}

fn main() {
    let mut builder = RowBuilder::new();
    builder.time.append_value(123);
    builder.obj1.append_value("123445");
    builder.many_name_value_pairs.keys().append_value("abc");
    builder.many_name_value_pairs.values().append_value(1);
    builder.many_name_value_pairs.keys().append_value("cds");
    builder.many_name_value_pairs.values().append_value(5848);
    builder.many_name_value_pairs.keys().append_value("edf");
    builder.many_name_value_pairs.values().append_value(749);
    builder.many_name_value_pairs.append(true).unwrap();
    let batch = builder.finish();
    print!("{}", pretty_format_batches(&[batch]).unwrap());
}
```

Should print

```
+-------------------------------+--------+-------------------------------+
| time                          | obj1   | many_name_value_pairs         |
+-------------------------------+--------+-------------------------------+
| 1970-01-01T00:00:00.000000123 | 123445 | {abc: 1, cds: 5848, edf: 749} |
+-------------------------------+--------+-------------------------------+
```

I opted to use MapBuilder as it appeared to match your schema best, however, you could also make use of ListBuilder and StructBuilder to store nested data differently. MapArray is effectively a special case of a ListArray of StructArray, where the StructArray has two children, one storing keys and one values.

**Reply by tghunt** [reply]
Hello,

Thank you for your help on this.  
I have taken the above and converted it to leverage ListBuilder and StructBuilder which can be seen in the following:
```
use std::sync::Arc;

use arrow::{
    array::{StructBuilder, Int32Builder, StringBuilder, ListBuilder},
    datatypes::{DataType, Field, Fields},
    record_batch::RecordBatch,
};

#[derive(Debug)]
struct RowBuilder {
    time: Int32Builder,
    obj1: StringBuilder,
    many_name_value_pairs: ListBuilder<StructBuilder>,
}

impl RowBuilder {
    fn new() -> Self {
        Self {
            time: Default::default(),
            obj1: Default::default(),
            many_name_value_pairs: ListBuilder::new(
                StructBuilder::new(
                    Fields::from(
                        vec![
                            Field::new("name", DataType::Utf8, true),
                            Field::new("value", DataType::Utf8, true),
                        ]
                    ),
                    vec![
                        Box::new(StringBuilder::new()),
                        Box::new(StringBuilder::new())
                    ]
                )
            )
        }
    }

    fn finish(&mut self) -> RecordBatch {
        RecordBatch::try_from_iter(vec![
            ("time", Arc::new(self.time.finish()) as _),
            ("obj1", Arc::new(self.obj1.finish()) as _),
            (
                "many_name_value_pairs",
                Arc::new(self.many_name_value_pairs.finish()) as _,
            ),
        ]).unwrap()
    }
}

fn main() {
    let mut builder = RowBuilder::new();
    builder.time.append_value(123);
    builder.obj1.append_value("123445");

    builder.many_name_value_pairs.values().field_builder::<StringBuilder>(0).unwrap().append_value("that");
    builder.many_name_value_pairs.values().field_builder::<StringBuilder>(1).unwrap().append_value("1a");
    builder.many_name_value_pairs.values().append(true);

    builder.many_name_value_pairs.values().field_builder::<StringBuilder>(0).unwrap().append_value("this");
    builder.many_name_value_pairs.values().field_builder::<StringBuilder>(1).unwrap().append_value("2b");
    builder.many_name_value_pairs.values().append(true);

    builder.many_name_value_pairs.values().field_builder::<StringBuilder>(0).unwrap().append_value("what");
    builder.many_name_value_pairs.values().field_builder::<StringBuilder>(1).unwrap().append_value("3c");
    builder.many_name_value_pairs.values().append(true);

    builder.many_name_value_pairs.append(true);

    let batch = builder.finish();

    println!("{:?}", batch);

}

```

However, in the above version and the mapbuilder version I end up with unnecessary nesting which can be seen in the following two images:
Using the MapBuider:
![mapper_table_4](https://github.com/apache/arrow-rs/assets/739514/ce64ef37-3d95-4757-afff-61abacf5c48b)

Using the ListBuilder and StructBuilder:
![listbuilder_structbuilder_table_4](https://github.com/apache/arrow-rs/assets/739514/536395f3-2378-4a0b-aebf-32f751136dfd)

Do you know of a way to eliminate the undesirable nesting in the record batch?  Both approaches seem to force nesting that I am not sure how to remove.

**Reply by tustvold** [reply]
There is no equivalent of a repeated struct in arrow, instead an equivalent construction can be composed from a list of structs. There is no semantic difference between the two representations, although the arrow version could also represent nullability at the group-level.

The parquet docs may be informative here, as parquet's schema is based on dremel - https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists

**Reply by tghunt** [reply]
To me it seems like the issue could potentially be resolved by updates to the recordbatch support.  If one could tell it to treat the struct of arrays as a single item, theoretically it could write out as it does in the example where the struct is imbedded in a list.  The record batch support currently does the opposite and thinks each object should have the same number of items.

**Reply by tustvold** [reply]
I'm not sure I follow what you mean, RecordBatch is a collection of columns, not a collection of rows?

**Reply by tghunt** [reply]
I think I am back to my original question about writing complex records to parquet.  I think the only way to accomplish it is to bypass any arrow support.  Is that even possible?

**Reply by tustvold** [reply]
Technically yes, practically you would need to perform the record shredding yourself, which is highly non-trivial. Is the additional nesting such a deal breaker for you? It is fairly standard practice AFAICT...

**Reply by tghunt** [reply]
Actually it is a deal breaker.  I know everything in Rust is still newish but I may have to revert back to Java (dislike its state of affairs) in order to build out the parquet files as needed.  As I have shown in the various images, the parquet files are ending up with arbitrary levels which in BigQuery means there are extra steps required to unnest data.  What should be a simple one level unnest to get at data becomes a multi-level unnest.  While it is not quite an issue from a software perspective, it is from a data science interactive perspective since the clean usability of the data is obscured with the arbitrary levels.

**Reply by tustvold** [reply]
My knowledge of BigQuery is limited, but as there is still the same number of repeated fields, I wouldn't have expected the arrow schema to require additional calls to unnest, which is only required for arrays, not non-repeated fields?

> I know everything in Rust is still newish

FWIW this isn't a Rust issue per-se, we are simply following the parquet standard in this regard :sweat_smile:

**Reply by tghunt** [reply]
With the avro crate I can generate the schema and layout out the data in the following manner:
![avro_schema_4](https://github.com/apache/arrow-rs/assets/739514/4b836e36-23ac-43db-9c13-5ecae68410e3)
![avro_sample_8](https://github.com/apache/arrow-rs/assets/739514/8f2a27bf-f649-49e7-a597-4559cc1bd759)

The unnesting is only required for the level of data/details you are after and the nested levels can be named appropriately.  

I was hoping to be able to write out a parquet file that would mimic this.  If it was not for the uncontrollable arbitrary items appearing, what we have explored would work.  In our examples I had many_name_value_pairs which would be tests in the image above.  With the Mapbuilder I would be able to rename entry to a test name, but if I have to store maps within a listbuilder I would end up with the extra list and items.

```
/*
toml adds:
apache-avro = "0.15.0"
serde = "1.0.163"
serde_json = "1.0.96"
*/
use std::fs::{File};
use apache_avro::{Schema, Writer};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct KeyValue {
    pub keyword: String,
    pub value: String
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Tests {
    test_name: String,
    test_details: Vec<KeyValue>
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Record {
    pub time: i64,
    pub id1: String,
    pub id2: String,
    pub tests: Vec<Tests>
}

pub fn main() {

    let mut record = Record{
        time: 123,
        id1: "id_1".to_string(),
        id2: "id_2".to_string(),
        tests: vec![],
    };

    for i in 0..5 {
        let test = format!("test_{}", i);

        let mut results = Vec::<KeyValue>::new();
        for i2 in 0..4 {
            results.push(KeyValue{ keyword: format!("result_{}", i2), value: format!("value_{}", i2) });
        }
        record.tests.push(Tests{ test_name: test, test_details: results })
    }
    let schema = get_schema();
    let serde_schema = Schema::parse_str(schema.as_str()).unwrap();
    let file = File::create("sample.avro").unwrap();
    let mut writer = Writer::new(&serde_schema, file);

    writer.append_ser(record).unwrap();
    writer.flush().expect("Flush to file failed");
}


/// Function that provides the table schema.
fn get_schema() -> String {
    let schema: String = r#"{
            "type": "record",
            "name": "Root",
            "fields": [
                {"name": "time", "type": "long", "logicalType": "timestamp"},
                {"name": "id1", "type": "string"},
                {"name": "id2", "type": "string"},
                {"name": "tests",
                    "type": {
                        "type": "array",
                        "items": {
                            "type": "record",
                            "namespace": "root.tests",
                            "name": "test",
                            "fields": [{
                                    "name": "test_name",
                                    "type": "string",
                                    "default": "null"
                                }, {
                                    "name": "test_details",
                                    "type": {
                                        "type": "array",
                                        "items": {
                                            "type": "record",
                                            "namespace": "root.tests.test_details",
                                            "name": "Test_Details",
                                            "fields": [{
                                                    "name": "keyword",
                                                    "type": "string",
                                                    "default": "null"
                                                }, {
                                                    "name": "value",
                                                    "type": "string",
                                                    "default": "null"
                                                }
                                            ]
                                        }
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }"#.to_string();
    schema
}

```

**Reply by tustvold** [reply]
>  If it was not for the uncontrollable arbitrary items appearing, what we have explored would work.

I am afraid I can't really offer you any further advice on this, other than to continue to strongly encourage you to follow the parquet schema conventions and adopt the list representation specified [here](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists), as this will ensure ecosystem interoperability. I would not expect there to be a material difference between querying the two schemas, as they will both have the same underlying dremel representation.

That being said if you are unable to follow the parquet conventions for schema design, you may want to email the parquet mailing list for advice on how best to achieve this, along with the long-term implications of such a design

**Comment by tghunt** [comment]
Is there a working example of this [code](https://gist.github.com/alamb/acd653c49e318ff70672b61325ba3443) somwhere?  I found the link from [this ](https://arrow.apache.org/blog/2022/10/17/arrow-parquet-encoding-part-3/) the last page of the arrow and parquet blog.

**Reply by tustvold** [reply]
Change

```
let b2 = DataType::List(Box::new(Field::new("element", DataType::Int32, false)));
```

To

```
let b2 = DataType::List(Arc::new(Field::new("element", DataType::Int32, false)));
```

And the array construction compiles

**Reply by tghunt** [reply]
Thank you.

I am not sure, but it appears the ListBuilder being assigned for b2 is changing the signature of b2 and the name "element" is being changed to "item".  If I change the name "element" to item and false to true to match what is in the struct it will run and I can print out the values object.

I was also able to get the values object out to a parquet via RecordBatch and again, there are arbitrary levels being added from the ListBuilder.

I am not seeing a way to get the structure of from the article using arrow objects.



![parquet_layout_from_article_4](https://github.com/apache/arrow-rs/assets/739514/d78346ba-e921-475f-a237-eeba35a40e01)

**Reply by tustvold** [reply]
You may be right, the parquet writer follows the specification and always maps a list to a 3 element structure, so as to correctly encode both list and element nullability. 

Whilst it supports reading 2 or 1 element lists for backwards compatibility, the arrow APIs will not write them
